Update readme (#856)
graphql-dotnet/server: README.md

> # ASP.NET Core Server for GraphQL.NET

Reviewer: ```suggestion
# ASP.NET Core GraphQL Server driven by GraphQL.NET
```

| GraphQL.Server.Ui.Voyager                            | [![Nuget](https://img.shields.io/nuget/dt/GraphQL.Server.Ui.Voyager)](https://www.nuget.org/packages/GraphQL.Server.Ui.Voyager)                                                       | [![Nuget](https://img.shields.io/nuget/v/GraphQL.Server.Ui.Voyager)](https://www.nuget.org/packages/GraphQL.Server.Ui.Voyager)                                                       | Provides Voyager UI middleware |

You can install the latest stable versions via [NuGet](https://www.nuget.org/packages/GraphQL.Server.Transports.AspNetCore/).
Also you can get all preview versions from [GitHub Packages](https://github.com/orgs/graphql-dotnet/packages?repo_name=server).
Note that GitHub requires authentication to consume the feed. See more information [here](https://docs.github.com/en/free-pro-team@latest/packages/publishing-and-managing-packages/about-github-packages#authenticating-to-github-packages).

## Description

> This package is designed for ASP.Net Core (2.1 through 6.0) to facilitate easy set-up of GraphQL requests

Reviewer: ```suggestion
This package is designed for ASP.NET Core (2.1 through 6.0) to facilitate easy set-up of GraphQL requests
```

over HTTP.  The code is designed to be used as middleware within the ASP.NET Core pipeline,
serving GET, POST or WebSocket requests.  GET requests process requests from the query string.
POST requests can be in the form of JSON requests, form submissions, or raw GraphQL strings.
WebSocket requests can use the `graphql-ws` or `graphql-transport-ws` WebSocket sub-protocol,
as defined in the [apollographql/subscriptions-transport-ws](https://github.com/apollographql/subscriptions-transport-ws)
and [enisdenjo/graphql-ws](https://github.com/enisdenjo/graphql-ws) respoitories, respectively.

> The `graphql-subscriptions` sub-protocol is not supported.

Reviewer: What is it?

Author: ```suggestion
```

I can't remember.  Maybe I read it wrong when reading some other notes.  Better just to delete this comment.

directives.  It skips validations for fields or fragments that are marked with the `@skip` or
`@include` directives.

See [migration notes](migration/migration7.md) for changes from version 6.x.

## Configuration

### Typical configuration with HTTP middleware

First add either the `GraphQL.Server.All` nuget package or the `GraphQL.Server.Transports.AspNetCore`
nuget package to your application.  Referencing the "all" package will include the UI middleware
packages.  These packages depend on `GraphQL` version 7.0.0 or later.

Then update your `Program.cs` or `Startup.cs` to configure GraphQL, registering the schema
and the serialization engine as a minimum.  Configure WebSockets and GraphQL in the HTTP
pipeline by calling `UseWebSockets` and `UseGraphQL` at the appropriate point.
Finally, you may also include some UI middleware for easy testing of your GraphQL endpoint
by calling `UseGraphQLVoyager` or a similar method at the appropriate point.

Below is a complete sample of a .NET 6 console app that hosts a GraphQL endpoint at
`http://localhost:5000/graphql`:

#### Project file

```xml
<Project Sdk="Microsoft.NET.Sdk.Web">

  <PropertyGroup>

>     <OutputType>Exe</OutputType>

Reviewer: ```suggestion
```

    <TargetFramework>net6.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="GraphQL.Server.All" Version="7.0.0" />
  </ItemGroup>

</Project>
```

#### Program.cs file

```csharp
using GraphQL;

var builder = WebApplication.CreateBuilder(args);
builder.Services.AddGraphQL(b => b
    .AddAutoSchema<Query>()  // schema
    .AddSystemTextJson());   // serializer

var app = builder.Build();
app.UseDeveloperExceptionPage();
app.UseWebSockets();
app.UseGraphQL("/graphql");            // url to host GraphQL endpoint
app.UseGraphQLPlayground(

>     new GraphQL.Server.Ui.Playground.PlaygroundOptions {

Reviewer: I would prefer to move opening curly brace on new line like it is across codebase.

>         GraphQLEndPoint = new PathString("/graphql"),         // url of GraphQL endpoint

Reviewer: TODO: change order and remove PathString

Author: This has been updated

        SubscriptionsEndPoint = new PathString("/graphql"),   // url of GraphQL endpoint
    },
    "/");                              // url to host Playground at
await app.RunAsync();
```

#### Schema

```csharp
public class Query
{
    public static string Hero() => "Luke Skywalker";
}
```

#### Sample request url

```
http://localhost:5000/graphql?query={hero}
```

#### Sample response

```json
{"data":{"hero":"Luke Skywalker"}}
```

### Configuration with endpoint routing

To use endpoint routing, call `MapGraphQL` from inside the endpoint configuration
builder rather than `UseGraphQL` on the application builder.  See below for the
sample of the application builder code:

```csharp
var app = builder.Build();
app.UseDeveloperExceptionPage();
app.UseWebSockets();
app.UseRouting();

> app.UseEndpoints(endpoints => {

Author: ```suggestion
app.UseEndpoints(endpoints =>
{
```

    public TestController(IDocumentExecuter<ISchema> documentExecuter)
    {
        _documentExecuter = documentExecuter;
    }

    [HttpGet]
    public async Task<IActionResult> GraphQL(string query)
    {
        var result = await _documentExecuter.ExecuteAsync(new() {
            Query = query,
            RequestServices = HttpContext.RequestServices,
            CancellationToken = HttpContext.RequestAborted,
        });
        return new ExecutionResultActionResult(result);
    }
}
```

### User context configuration

To set the user context to be used during the execution of GraphQL requests,
call `AddUserContextBuilder` during the GraphQL service setup to set a delegate
which will be called when the user context is built.  Alternatively, you can
register an `IUserContextBuilder` implementation to do the same.

#### Program.cs / Startup.cs

```csharp
builder.Services.AddGraphQL(b => b
    .AddAutoSchema<Query>()
    .AddSystemTextJson()
    .AddUserContextBuilder(httpContext => new MyUserContext(httpContext));
```

#### MyUserContext.cs

```csharp
public class MyUserContext : Dictionary<string, object?>
{
    public ClaimsPrincipal User { get; }

>     public MyUserContext(HttpContext context) : base()

Reviewer: ```suggestion
    public MyUserContext(HttpContext context)
```

#### Authentication for WebSocket requests

Since WebSocket requests from browsers cannot typically carry a HTTP Authorization header, you
will need to authorize requests via the `ConnectionInit` WebSocket message or carry the authorization
token within the URL.  Below is a sample of the former:

```cs
builder.Services.AddGraphQL(b => b
    .AddAutoSchema<Query>()
    .AddSystemTextJson()
    .AddAuthorizationRule()  // not required for endpoint authorization
    .AddWebSocketAuthentication<MyAuthService>());

app.UseGraphQL("/graphql", config =>
{
    // require that the user be authenticated
    config.AuthorizationRequired = true;
});

class MyAuthService : IWebSocketAuthenticationService
{
    private readonly IGraphQLSerializer _serializer;

    public MyWSAuthService(IGraphQLSerializer serializer)
    {
        _serializer = serializer;
    }

    public async ValueTask<bool> AuthenticateAsync(IWebSocketConnection connection, OperationMessage operationMessage)
    {
        // read payload of ConnectionInit message and look for an "Authorization" entry that starts with "Bearer "
        var payload = _serializer.ReadNode<Inputs>(operationMessage.Payload);

>         if (payload?.TryGetValue("Authorization", out var value) ?? false)

Author: ```suggestion
        if ((payload?.TryGetValue("Authorization", out var value) ?? false) && value is string valueString)
```

        {

>             var user = ParseToken(value);

Reviewer: value is `object`

Author: ```suggestion
            var user = ParseToken(valueString);
```

disabling `HandleGet` or similar; `HandleAuthorizeAsync` and `HandleAuthorizeWebSocketConnectionAsync`
will not.

### UI configuration

There are four UI middleware projects included; Altair, GraphiQL, Playground and Voyager.
See review the following methods for configuration options within each of the 4 respective
NuGet packages:

```csharp
app.UseGraphQLAltair();
app.UseGraphQLGraphiQL();
app.UseGraphQLPlayground();
app.UseGraphQLVoyager();

// or

endpoints.MapGraphQLAltair();
endpoints.MapGraphQLGraphiQL();
endpoints.MapGraphQLPlayground();
endpoints.MapGraphQLVoyager();
```

### CORS configuration

ASP.NET Core supports CORS requests independently of GraphQL, including CORS pre-flight
requests.  To configure your application for CORS requests, add `AddCors()` and `UseCors()`
into the application pipeline.

```csharp
builder.Services.AddCors();

app.UseCors(policy => {
    // configure default policy here
});
```

To configure GraphQL to use a named CORS policy, configure the application to use endpoint routing
and call `RequireCors()` on the endpoint configuration builder.

```csharp
// ...
builder.Services.AddRouting();

> builder.Services.AddCors(builder => {

Reviewer: ```suggestion
builder.Services.AddCors(builder =>
{
```

ASP.NET Core supports response compression independently of GraphQL, with brotli and gzip
support automatically based on the compression formats listed as supported in the request headers.
To configure your application for response compression, configure your Program/Startup file as
follows:

```csharp
// add and configure the service
builder.Services.AddResponseCompression(options => {
    options.EnableForHttps = true; // may lead to CRIME and BREACH attacks
    options.MimeTypes = new[] { "application/json", "application/graphql+json" };
})

// place this first/early in the pipeline
app.UseResponseCompression();
```

In order to compress GraphQL responses, the `application/graphql+json` content type must be
added to the `MimeTypes` option.  You may choose to enable other content types as well.

Please note that enabling response compression over HTTPS can lead to CRIME and BREACH
attacks.  These side-channel attacks typically affects sites that rely on cookies for
authentication.  Please read [this](https://docs.microsoft.com/en-us/aspnet/core/performance/response-compression?view=aspnetcore-6.0)
and [this](http://www.breachattack.com/#howitworks) for more details.

> ### ASP.NET Core 2.1

Reviewer: ```suggestion
### ASP.NET Core 2.1 / .NET Framework 4.8
```

The only additional requirement is that you must add this code in your `Startup.cs` file:

```csharp
services.AddHostApplicationLifetime();
```

Besides that requirement, all features are supported in exactly the same manner as
when using ASP.NET Core 3.1+.  You may find differences in the ASP.NET Core runtime,
such as CORS implementation differences, which are outside the scope of this project.

Please note that a serializer reference is not included for these projects within
`GraphQL.Server.Transports.AspNetCore`; you will need to reference either
`GraphQL.NewtonsoftJson` or `GraphQL.SystemTextJson`, or reference
`GraphQL.Server.All` which includes `GraphQL.NewtonsoftJson` for ASP.NET Core 2.1 projects.
This is because `Newtonsoft.Json` is the default serializer for ASP.NET Core 2.1
rather `System.Text.Json`.  When using `GraphQL.NewtonsoftJson`, you will need to call
`AddNewtonsoftJson()` rather than `AddSystemTextJson()` while configuring GraphQL.NET.

## Advanced configuration

For more advanced configurations, see the overloads and configuration options
available for the various builder methods, listed below.  Methods and properties

> contain XML comments to provide assistance while coding with Visual Studio.

Reviewer: ```suggestion
contain XML comments to provide assistance while coding with your IDE.
```

| `KeepAliveTimeout`          | The amount of time to wait between sending keep-alive packets. | disabled |
| `DisconnectionTimeout`      | The amount of time to wait to attempt a graceful teardown of the WebSockets protocol. | 10 seconds |
| `DisconnectAfterErrorEvent` | Disconnects a subscription from the client if the subscription source dispatches an `OnError` event. | True |
| `DisconnectAfterAnyError`   | Disconnects a subscription from the client there are any GraphQL errors during a subscription. | False |

### Multi-schema configuration

You may use the generic versions of the various builder methods to map a URL to a particular schema.

```csharp
var app = builder.Build();
app.UseDeveloperExceptionPage();
app.UseGraphQL<DogSchema>("/dogs/graphql");
app.UseGraphQL<CatSchema>("/cats/graphql");
await app.RunAsync();
```

### Different global authorization settings for different transports (GET/POST/WebSockets)

You may register the same endpoint multiple times if necessary to configure GET connections
with certain authorization options, and POST connections with other authorization options.

```csharp
var app = builder.Build();
app.UseDeveloperExceptionPage();

> app.UseGraphQL("/graphql", options => {

Reviewer: ```suggestion
app.UseGraphQL("/graphql", options =>
{
```

    options.HandleGet = true;
    options.HandlePost = false;
    options.HandleWebSockets = false;
    options.AuthorizationRequired = false;
});

> app.UseGraphQL("/graphql", options => {

Reviewer: ```suggestion
app.UseGraphQL("/graphql", options =>
{
```

    options.HandleGet = false;
    options.HandlePost = true;
    options.HandleWebSockets = true;
    options.AuthorizationRequired = true;   // require authentication for POST/WebSocket connections
});
await app.RunAsync();
```

Since POST and WebSockets can be used for query requests, it is recommended not to do the above,
but instead add the authorization validation rule and add authorization metadata on the Mutation
and Subscription portions of your schema, as shown below:

```csharp
builder.Services.AddGraphQL(b => b
    .AddSchema<MySchema>()
    .AddSystemTextJson()

>     .AddAuthorization()); // add authorization validation rule

Author: ```suggestion
    .AddAuthorizationRule()); // add authorization validation rule
```

var app = builder.Build();
app.UseDeveloperExceptionPage();
app.UseGraphQL();
await app.RunAsync();

// demonstration of code-first schema; also possible with schema-first or type-first schemas
public class MySchema : Schema
{
    public MySchema(IServiceProvider provider, MyQuery query, MyMutation mutation) : base(provider)
    {
        Query = query;
        Mutation = mutation;

        mutation.Authorize(); // require authorization for any mutation request
    }
}
```

### Customizing middleware behavior

GET/POST requests are handled directly by the `GraphQLHttpMiddleware`.

> For WebSocket requests an `WebSocketConnection` instance is instated to dispatch incoming

Reviewer: ```suggestion
For WebSocket requests an `WebSocketConnection` instance is instantiated to dispatch incoming
```

messages and send outgoing messages.  Depending on the WebSocket sub-protocols supported by the

> client, the proper implementation of `IOperationMessageProcessor` is instated to act as a

Reviewer: ```suggestion
client, the proper implementation of `IOperationMessageProcessor` is created to act as a
```

| `HandleWebSocketSubProtocolNotSupportedAsync` | Writes a '400 Invalid WebSocket sub-protocol.' message to the output. |

Below is a sample of custom middleware to change the response content type to `application/json`,
rather than the default of `application/graphql+json`:

```csharp
class MyMiddleware<TSchema> : GraphQLHttpMiddleware<TSchema>
    where TSchema : ISchema
{
    public MyMiddleware(
        RequestDelegate next,
        IGraphQLTextSerializer serializer,
        IDocumentExecuter<TSchema> documentExecuter,
        IServiceScopeFactory serviceScopeFactory,
        GraphQLHttpMiddlewareOptions options,
        IHostApplicationLifetime hostApplicationLifetime)
        : base(next, serializer, documentExecuter, serviceScopeFactory, options, hostApplicationLifetime)
    {
    }

    protected override string SelectResponseContentType(HttpContext context)
        => "application/json";
}

app.UseGraphQL<MyMiddleware<ISchema>>("/graphql", new GraphQLHttpMiddlewareOptions());
```

#### WebSocket handler classes

The WebSocket handling code is organized as follows:

| Interface / Class                | Description |
|----------------------------------|-------------|

> | `IWebSocketConnection`           | Provides methods to a send a message to a client or close the connection. |

Reviewer: ```suggestion
| `IWebSocketConnection`           | Provides methods to send a message to a client or close the connection. |
```

| Basic           | .NET 6 Minimal           | Demonstrates simplest possible implementation |
| Complex         | .NET 3.1 / 5 / 6         | Demonstrates older Program/Startup files and various configuration options, and multiple UI endpoints |
| Controller      | .NET 6 Minimal           | MVC implementation; does not include WebSocket support |
| Cors            | .NET 6 Minimal           | Demonstrates configuring a GraphQL endpoint to use a specified CORS policy |
| EndpointRouting | .NET 6 Minimal           | Demonstrates configuring GraphQL through endpoint routing |
| MultipleSchemas | .NET 6 Minimal           | Demonstrates configuring multiple schemas within a single server |
| Net48           | .NET Core 2.1 / .NET 4.8 | Demonstrates configuring GraphQL on .NET 4.8 / Core 2.1 |
| Pages           | .NET 6 Minimal           | Demonstrates configuring GraphQL on top of a Razor Pages template |

Most of the above samples rely on a sample "Chat" schema.
Below are some basic requests you can use to test the schema:

### Queries

#### Return number of messages

> ```gql

Reviewer: ```suggestion
```graphql
```

[![Run code tests](https://github.com/graphql-dotnet/server/actions/workflows/test.yml/badge.svg)](https://github.com/graphql-dotnet/server/actions/workflows/test.yml)
[![Build artifacts](https://github.com/graphql-dotnet/server/actions/workflows/build.yml/badge.svg)](https://github.com/graphql-dotnet/server/actions/workflows/build.yml)
[![Publish release](https://github.com/graphql-dotnet/server/actions/workflows/publish.yml/badge.svg)](https://github.com/graphql-dotnet/server/actions/workflows/publish.yml)
[![CodeQL analysis](https://github.com/graphql-dotnet/server/actions/workflows/codeql-analysis.yml/badge.svg)](https://github.com/graphql-dotnet/server/actions/workflows/codeql-analysis.yml)

[![codecov](https://codecov.io/gh/graphql-dotnet/server/branch/master/graph/badge.svg?token=ZBcVYq7hz4)](https://codecov.io/gh/graphql-dotnet/server)

![Activity](https://img.shields.io/github/commit-activity/w/graphql-dotnet/server)
![Activity](https://img.shields.io/github/commit-activity/m/graphql-dotnet/server)

> ![Activity](https://img.shields.io/github/commit-activity/y/graphql-dotnet/server)

Reviewer: ?

Author: It’s pointless garbage at the top of the repository. Who cares how many commits there are in the last week, month or year? The readme didn’t start until halfway down the first page before. Let’s keep whatever of these kinds of tags fit on a single line. I put the version, coverage, and downloads, I think (?). The version is important to returning users. Coverage and downloads help indicate how supported and popular it is.

Author: I also put the license and repo size. License is important to commercial users. Repo size is probably rather pointless also.

Reviewer: >  Who cares how many commits there are in the last week, month or year? 

You answered yourself : 

> Coverage and downloads help indicate **how supported** and popular it is.

Author: Okay, but this project has historically had low commits.  It does not benefit the project to show small numbers.  Let's show "4.7 million downloads" and "95% coverage", not "1 commit last week".

Author: This is not impressive for a project with 4.7m downloads; makes it look more like the project is dead:
![image](https://user-images.githubusercontent.com/6377684/178587652-db4b48d8-f21c-4333-9db0-b218c8420458.png)

Reviewer: It's because of `master` branch. Try https://img.shields.io/github/commit-activity/w/graphql-dotnet/server/develop but I understand your thoughts. What about https://img.shields.io/github/contributors/graphql-dotnet/server ? Maybe you'll find something usefull here https://shields.io/category/activity ?

Author: I'm in no hurry to add more tags.  If you like, pick one more and we will add it.  Contributors is fine.

Maybe latest release date:

https://img.shields.io/github/release-date/graphql-dotnet/server?label=latest%20release

Or maybe commits since latest release:

https://img.shields.io/github/commits-since/graphql-dotnet/server/latest

Those last two seem that they might be useful for users of the library.

Reviewer: Agreem let's add them.
